common/src/main/java/org/conscrypt/OAEPParameters.java

```diff
@@ -0,0 +1,272 @@
+ * limitations under the License.
+ */
+
+package org.conscrypt;
+
+import java.io.IOException;
+import java.security.AlgorithmParametersSpi;
+import java.security.spec.AlgorithmParameterSpec;
+import java.security.spec.InvalidParameterSpecException;
+import java.security.spec.MGF1ParameterSpec;
+import java.util.HashMap;
+import java.util.Map;
+import javax.crypto.spec.OAEPParameterSpec;
+import javax.crypto.spec.PSource;
+
+/**
+ * AlgorithmParameters implementation for OAEP.  The only supported encoding format is ASN.1,
+ * as specified in RFC 4055 section 4.1.
+ *
+ * @hide
+ */
+@Internal
+public class OAEPParameters extends AlgorithmParametersSpi {
+
+    private static final Map<String, String> OID_TO_NAME = new HashMap<>();
+    private static final Map<String, String> NAME_TO_OID = new HashMap<>();
+    static {
+        OID_TO_NAME.put("1.3.14.3.2.26", "SHA-1");
+        OID_TO_NAME.put("2.16.840.1.101.3.4.2.4", "SHA-224");
+        OID_TO_NAME.put("2.16.840.1.101.3.4.2.1", "SHA-256");
+        OID_TO_NAME.put("2.16.840.1.101.3.4.2.2", "SHA-384");
+        OID_TO_NAME.put("2.16.840.1.101.3.4.2.3", "SHA-512");
+        for (Map.Entry<String, String> entry : OID_TO_NAME.entrySet()) {
+            NAME_TO_OID.put(entry.getValue(), entry.getKey());
+        }
+    }
+    private static final String MGF1_OID = "1.2.840.113549.1.1.8";
+    private static final String PSPECIFIED_OID = "1.2.840.113549.1.1.9";
+
+    private OAEPParameterSpec spec = OAEPParameterSpec.DEFAULT;
+
+    @Override
+    protected void engineInit(AlgorithmParameterSpec algorithmParameterSpec)
+            throws InvalidParameterSpecException {
+        if (algorithmParameterSpec instanceof OAEPParameterSpec) {
+            this.spec = (OAEPParameterSpec) algorithmParameterSpec;
+        } else {
+            throw new InvalidParameterSpecException("Only OAEPParameterSpec is supported");
+        }
+    }
+
+    @Override
+    protected void engineInit(byte[] bytes) throws IOException {
+        long readRef = 0;
+        long seqRef = 0;
+        try {
+            readRef = NativeCrypto.asn1_read_init(bytes);
+            seqRef = NativeCrypto.asn1_read_sequence(readRef);
+            String hash = "SHA-1";
+            String mgfHash = "SHA-1";
+            PSource.PSpecified pSpecified = PSource.PSpecified.DEFAULT;
+            if (NativeCrypto.asn1_read_next_tag_is(seqRef, 0)) {
+                long hashRef = 0;
+                try {
+                    hashRef = NativeCrypto.asn1_read_tagged(seqRef);
+                    hash = getHashName(hashRef);
+                } finally {
+                    NativeCrypto.asn1_read_free(hashRef);
+                }
+            }
+            if (NativeCrypto.asn1_read_next_tag_is(seqRef, 1)) {
+                long mgfRef = 0;
+                long mgfSeqRef = 0;
+                try {
+                    mgfRef = NativeCrypto.asn1_read_tagged(seqRef);
```

Comment:
I considered that, but the sequence needs to return two values (the OID and the reference to read the params from), and I thought the reduction in clarity from having to pass back multiple values outweighed not repeating the code.